// @sequelize/core/lib/errors/database-error.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var database_error_exports = {};
__export(database_error_exports, {
  default: () => database_error_default
});
var import_base_error = __toESM(require("./base-error"));
class DatabaseError extends import_base_error.default {
  parent;
  original;
  sql;
  parameters;
  constructor(parent, options = {}) {
    super(parent.message);
    var _a;
    this.name = "SequelizeDatabaseError";
    this.parent = parent;
    this.original = parent;
    this.sql = parent.sql;
    this.parameters = (_a = parent.parameters) != null ? _a : {};
    if (options.stack) {
      this.stack = options.stack;
    }
  }
}
var database_error_default = DatabaseError;
module.exports = __toCommonJS(database_error_exports);
