// @sequelize/core/lib/query-types.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var query_types_exports = {};
__export(query_types_exports, {
  QueryTypes: () => QueryTypes
});
var QueryTypes = /* @__PURE__ */ ((QueryTypes2) => {
  QueryTypes2["SELECT"] = "SELECT";
  QueryTypes2["INSERT"] = "INSERT";
  QueryTypes2["UPDATE"] = "UPDATE";
  QueryTypes2["BULKUPDATE"] = "BULKUPDATE";
  QueryTypes2["BULKDELETE"] = "BULKDELETE";
  QueryTypes2["DELETE"] = "DELETE";
  QueryTypes2["UPSERT"] = "UPSERT";
  QueryTypes2["VERSION"] = "VERSION";
  QueryTypes2["SHOWTABLES"] = "SHOWTABLES";
  QueryTypes2["SHOWINDEXES"] = "SHOWINDEXES";
  QueryTypes2["DESCRIBE"] = "DESCRIBE";
  QueryTypes2["RAW"] = "RAW";
  QueryTypes2["FOREIGNKEYS"] = "FOREIGNKEYS";
  QueryTypes2["SHOWCONSTRAINTS"] = "SHOWCONSTRAINTS";
  return QueryTypes2;
})(QueryTypes || {});
module.exports = __toCommonJS(query_types_exports);
