// @sequelize/core/lib/table-hints.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var table_hints_exports = {};
__export(table_hints_exports, {
  TableHints: () => TableHints
});
var TableHints = /* @__PURE__ */ ((TableHints2) => {
  TableHints2["NOLOCK"] = "NOLOCK";
  TableHints2["READUNCOMMITTED"] = "READUNCOMMITTED";
  TableHints2["UPDLOCK"] = "UPDLOCK";
  TableHints2["REPEATABLEREAD"] = "REPEATABLEREAD";
  TableHints2["SERIALIZABLE"] = "SERIALIZABLE";
  TableHints2["READCOMMITTED"] = "READCOMMITTED";
  TableHints2["TABLOCK"] = "TABLOCK";
  TableHints2["TABLOCKX"] = "TABLOCKX";
  TableHints2["PAGLOCK"] = "PAGLOCK";
  TableHints2["ROWLOCK"] = "ROWLOCK";
  TableHints2["NOWAIT"] = "NOWAIT";
  TableHints2["READPAST"] = "READPAST";
  TableHints2["XLOCK"] = "XLOCK";
  TableHints2["SNAPSHOT"] = "SNAPSHOT";
  TableHints2["NOEXPAND"] = "NOEXPAND";
  return TableHints2;
})(TableHints || {});
module.exports = __toCommonJS(table_hints_exports);
