// @sequelize/core/lib/errors/validation/unique-constraint-error.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var unique_constraint_error_exports = {};
__export(unique_constraint_error_exports, {
  default: () => unique_constraint_error_default
});
var import_validation_error = __toESM(require("../validation-error"));
class UniqueConstraintError extends import_validation_error.default {
  parent;
  original;
  fields;
  sql;
  constructor(options) {
    var _a, _b, _c;
    options = options != null ? options : {};
    options.parent = (_a = options.parent) != null ? _a : { sql: "", name: "", message: "" };
    options.message = options.message || options.parent.message || "Validation Error";
    options.errors = (_b = options.errors) != null ? _b : [];
    super(options.message, options.errors, { stack: options.stack });
    this.name = "SequelizeUniqueConstraintError";
    this.fields = (_c = options.fields) != null ? _c : {};
    this.parent = options.parent;
    this.original = options.parent;
    this.sql = options.parent.sql;
  }
}
var unique_constraint_error_default = UniqueConstraintError;
module.exports = __toCommonJS(unique_constraint_error_exports);
